File: apps/yukino/src/todo/tools.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { safeParseAsync } from "@modelcontextprotocol/sdk/server/zod-compat.js";
import z from "zod";

import type { StoredTaskStatus } from "./store.js";

import type { TaskList } from "./index.js";

import type {
  Tool,
  ToolResult,
  ToolContext,
  ToolSchema,
} from "@/tools/types.js";
import { asErrorString, strArg } from "@/utils/index.js";

export class TaskCreateTool implements Tool {
  name = "TaskCreate";
  description = "Create a new task to track work.";
  category = "read" as const;

  private list: TaskList;

  constructor(list: TaskList) {
    this.list = list;
  }

  schema(): ToolSchema {
    return {
      name: this.name,
      description: this.description,
      input_schema: {
        type: "object",
        properties: {
          subject: { type: "string", description: "Brief task title" },
          description: { type: "string", description: "What needs to be done" },
          activeForm: {
            type: "string",
            description: "Present continuous form for spinner",
          },
        },
        required: ["subject", "description"],
      },
    };
  }

  execute(
    ctx: ToolContext,
    args: Record<string, unknown>,
  ): Promise<ToolResult> {
    const subject = strArg(args, "subject");
    const description = strArg(args, "description");
    const activeForm = strArg(args, "activeForm") || undefined;
    if (!subject) {
      return Promise.resolve({
        output: "Error: subject is required",
        isError: true,
      });
    }
    const task = this.list.create(subject, description, activeForm);
    return Promise.resolve({
      output: `Task #${task.id} created successfully: ${task.subject}`,
      isError: false,
    });
  }
}

export class TaskGetTool implements Tool {
  name = "TaskGet";
  description = "Get a task by its ID.";
  category = "read" as const;

  private list: TaskList;

  constructor(list: TaskList) {
    this.list = list;
  }

  schema(): ToolSchema {
    return {
      name: this.name,
      description: this.description,
      input_schema: {
        type: "object",
        properties: { taskId: { type: "string", description: "Task ID" } },
        required: ["taskId"],
      },
    };
  }

  execute(
    ctx: ToolContext,
    args: Record<string, unknown>,
  ): Promise<ToolResult> {
    const id = strArg(args, "taskId");
    const task = this.list.get(id);
    if (!task) {
      return Promise.resolve({ output: "Task not found", isError: true });
    }
    return Promise.resolve({
      output: JSON.stringify(task, null, 2),
      isError: false,
    });
  }
}

export class TaskListTool implements Tool {
  name = "TaskList";
  description = "List all tasks.";
  category = "read" as const;

  private list: TaskList;

  constructor(list: TaskList) {
    this.list = list;
  }

  schema(): ToolSchema {
    return {
      name: this.name,
      description: this.description,
      input_schema: { type: "object", properties: {} },
    };
  }

  execute(): Promise<ToolResult> {
    const tasks = this.list.list();
    if (tasks.length === 0) {
      return Promise.resolve({ output: "No tasks found", isError: false });
    }
    const lines = tasks.map(
      (t) =>
        `#${t.id}. [${t.status}] ${t.subject}${t.owner ? ` (${t.owner})` : ""}`,
    );
    return Promise.resolve({ output: lines.join("\n"), isError: false });
  }
}

export class TaskUpdateTool implements Tool {
  name = "TaskUpdate";
  description = "Update a task's status, subject, or other fields.";
  category = "read" as const;

  private list: TaskList;

  constructor(list: TaskList) {
    this.list = list;
  }

  schema(): ToolSchema {
    return {
      name: this.name,
      description: this.description,
      input_schema: {
        type: "object",
        properties: {
          taskId: { type: "string", description: "Task ID" },
          status: {
            type: "string",
            description: "New status: pending, in_progress, completed, deleted",
          },
          subject: { type: "string", description: "New subject" },
          description: { type: "string", description: "New description" },
          owner: { type: "string", description: "New owner" },
          addBlocks: {
            type: "array",
            items: { type: "string" },
            description: "Tasks this one blocks",
          },
          addBlockedBy: {
            type: "array",
            items: { type: "string" },
            description: "Tasks blocking this one",
          },
        },
        required: ["taskId"],
      },
    };
  }

  async execute(
    ctx: ToolContext,
    args: Record<string, unknown>,
  ): Promise<ToolResult> {
    const result = await safeParseAsync(TaskUpdateArgsSchema, args);
    if (!result.success) {
      return {
        output: asErrorString(result.error),
        isError: true,
      };
    }

    const {
      taskId,
      status,
      subject,
      description,
      owner,
      addBlocks,
      addBlockedBy,
    } = result.data;

    if (!taskId) {
      return Promise.resolve({
        output: "Error: taskId is required",
        isError: true,
      });
    }

    if (status === "deleted") {
      this.list.delete(taskId);
      return Promise.resolve({
        output: `Task #${taskId} deleted`,
        isError: false,
      });
    }

    type Updates = Omit<TaskUpdateArgs, "taskId" | "status"> & {
      status?: StoredTaskStatus;
    };

    const updates: Updates = {};

    if (status) {
      updates.status = status;
    }
    if (subject) {
      updates.subject = subject;
    }
    if (description) {
      updates.description = description;
    }
    if (owner) {
      updates.owner = owner;
    }
    const task = this.list.update(taskId, updates);
    if (!task) {
      return { output: "Task not found", isError: true };
    }
    if (addBlocks) {
      this.list.addBlocks(taskId, addBlocks);
    }
    if (addBlockedBy) {
      this.list.addBlockedBy(taskId, addBlockedBy);
    }

    return { output: `Updated task #${taskId} status`, isError: false };
  }
}

const TaskUpdateArgsSchema = z.object({
  taskId: z.string(),
  subject: z.string().optional(),
  description: z.string().optional(),
  status: z.enum(["pending", "in_progress", "completed", "deleted"]).optional(),
  owner: z.string().optional(),
  addBlocks: z.array(z.string()).optional(),
  addBlockedBy: z.array(z.string()).optional(),
});

type TaskUpdateArgs = z.infer<typeof TaskUpdateArgsSchema>;

File: yukino/todo/tools.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package todo

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/hangtiancheng/yukino-code/yukino/tools"
)

// TaskCreateTool creates a new task in the todo list.
type TaskCreateTool struct {
	List *TaskList
}

func (t *TaskCreateTool) Name() string { return "TaskCreate" }

// CategoryRead matches the TS tool's `category = "read"` (tools.ts:41).
func (t *TaskCreateTool) Category() tools.ToolCategory { return tools.CategoryRead }

func (t *TaskCreateTool) Description() string {
	return "Create a new task to track work."
}

func (t *TaskCreateTool) Schema() map[string]any {
	return map[string]any{
		"name":        t.Name(),
		"description": t.Description(),
		"input_schema": map[string]any{
			"type": "object",
			"properties": map[string]any{
				"subject":     map[string]any{"type": "string", "description": "Brief task title"},
				"description": map[string]any{"type": "string", "description": "What needs to be done"},
				"activeForm": map[string]any{
					"type":        "string",
					"description": "Present continuous form for spinner",
				},
				// Go extension (TS Task carries metadata but the TS tool schema
				// does not expose it).
				"metadata": map[string]any{
					"type":        "object",
					"description": "Arbitrary metadata to attach to the task",
				},
			},
			"required": []string{"subject", "description"},
		},
	}
}

func (t *TaskCreateTool) Execute(_ context.Context, args map[string]any) tools.ToolResult {
	subject, _ := args["subject"].(string)
	// TS execute() only enforces subject; description rides through as-is
	// (tools.ts:72-81).
	if subject == "" {
		return tools.ToolResult{Output: "Error: subject is required", IsError: true}
	}
	description, _ := args["description"].(string)

	activeForm, _ := args["activeForm"].(string)
	metadata, _ := args["metadata"].(map[string]any)

	task, err := t.List.Create(subject, description, activeForm, metadata)
	if err != nil {
		return tools.ToolResult{Output: fmt.Sprintf("Error creating task: %s", err), IsError: true}
	}

	return tools.ToolResult{Output: fmt.Sprintf("Task #%s created successfully: %s", task.ID, task.Subject)}
}

// TaskGetTool retrieves a task by ID.
type TaskGetTool struct {
	List *TaskList
}

func (t *TaskGetTool) Name() string                 { return "TaskGet" }
func (t *TaskGetTool) Category() tools.ToolCategory { return tools.CategoryRead }

func (t *TaskGetTool) Description() string {
	return "Get a task by its ID."
}

func (t *TaskGetTool) Schema() map[string]any {
	return map[string]any{
		"name":        t.Name(),
		"description": t.Description(),
		"input_schema": map[string]any{
			"type": "object",
			"properties": map[string]any{
				"taskId": map[string]any{"type": "string", "description": "Task ID"},
			},
			"required": []string{"taskId"},
		},
	}
}

func (t *TaskGetTool) Execute(_ context.Context, args map[string]any) tools.ToolResult {
	taskID, _ := args["taskId"].(string)

	task, err := t.List.Get(taskID)
	if err != nil {
		return tools.ToolResult{Output: fmt.Sprintf("Error: %s", err), IsError: true}
	}
	if task == nil {
		return tools.ToolResult{Output: "Task not found", IsError: true}
	}

	// TS returns the task as pretty-printed JSON (tools.ts:121-124).
	data, err := json.MarshalIndent(task, "", "  ")
	if err != nil {
		return tools.ToolResult{Output: fmt.Sprintf("Error: %s", err), IsError: true}
	}
	return tools.ToolResult{Output: string(data)}
}

// TaskListTool lists all tasks.
type TaskListTool struct {
	List *TaskList
}

func (t *TaskListTool) Name() string                 { return "TaskList" }
func (t *TaskListTool) Category() tools.ToolCategory { return tools.CategoryRead }

func (t *TaskListTool) Description() string {
	return "List all tasks."
}

func (t *TaskListTool) Schema() map[string]any {
	return map[string]any{
		"name":        t.Name(),
		"description": t.Description(),
		"input_schema": map[string]any{
			"type":       "object",
			"properties": map[string]any{},
		},
	}
}

func (t *TaskListTool) Execute(_ context.Context, _ map[string]any) tools.ToolResult {
	tasks, err := t.List.List()
	if err != nil {
		return tools.ToolResult{Output: fmt.Sprintf("Error: %s", err), IsError: true}
	}
	if len(tasks) == 0 {
		return tools.ToolResult{Output: "No tasks found"}
	}

	// TS renders `#id. [status] subject (owner)` and nothing else
	// (tools.ts:152-156); the dependency links stay in the data model.
	lines := make([]string, 0, len(tasks))
	for _, task := range tasks {
		line := fmt.Sprintf("#%s. [%s] %s", task.ID, task.Status, task.Subject)
		if task.Owner != "" {
			line += fmt.Sprintf(" (%s)", task.Owner)
		}
		lines = append(lines, line)
	}
	return tools.ToolResult{Output: strings.Join(lines, "\n")}
}

// TaskUpdateTool updates an existing task.
type TaskUpdateTool struct {
	List *TaskList
}

func (t *TaskUpdateTool) Name() string { return "TaskUpdate" }

// CategoryRead matches the TS tool's `category = "read"` (tools.ts:163).
func (t *TaskUpdateTool) Category() tools.ToolCategory { return tools.CategoryRead }

func (t *TaskUpdateTool) Description() string {
	return "Update a task's status, subject, or other fields."
}

func (t *TaskUpdateTool) Schema() map[string]any {
	return map[string]any{
		"name":        t.Name(),
		"description": t.Description(),
		"input_schema": map[string]any{
			"type": "object",
			"properties": map[string]any{
				"taskId":  map[string]any{"type": "string", "description": "Task ID"},
				"subject": map[string]any{"type": "string", "description": "New subject"},
				// TS spells the status enum through zod validation at execute
				// time, not in the schema (tools.ts:179-182,276).
				"status": map[string]any{
					"type":        "string",
					"description": "New status: pending, in_progress, completed, deleted",
				},
				"description": map[string]any{"type": "string", "description": "New description"},
				"owner":       map[string]any{"type": "string", "description": "New owner"},
				"addBlocks": map[string]any{
					"type":        "array",
					"items":       map[string]any{"type": "string"},
					"description": "Tasks this one blocks",
				},
				"addBlockedBy": map[string]any{
					"type":        "array",
					"items":       map[string]any{"type": "string"},
					"description": "Tasks blocking this one",
				},
				// Go extension (documented): mergeable metadata with
				// null-deletes-key semantics.
				"metadata": map[string]any{
					"type":        "object",
					"description": "Metadata keys to merge. Set a key to null to delete it.",
				},
			},
			"required": []string{"taskId"},
		},
	}
}

func (t *TaskUpdateTool) Execute(_ context.Context, args map[string]any) tools.ToolResult {
	taskID, _ := args["taskId"].(string)
	if taskID == "" {
		return tools.ToolResult{Output: "Error: taskId is required", IsError: true}
	}

	// Reject invalid statuses up front, mirroring the TS zod enum parse
	// (tools.ts:206-212,276). The TS failure text is the serialized zod issue
	// list, which has no stable Go equivalent; the Go wording is a documented
	// migration difference.
	status, hasStatus := args["status"].(string)
	if hasStatus && !validUpdateStatus(status) {
		return tools.ToolResult{
			Output:  fmt.Sprintf("Error: invalid status %v: must be one of pending, in_progress, completed, deleted", args["status"]),
			IsError: true,
		}
	}

	// TS handles the delete before any update and unconditionally reports
	// success (tools.ts:231-237).
	if hasStatus && status == "deleted" {
		_, _ = t.List.Delete(taskID)
		return tools.ToolResult{Output: fmt.Sprintf("Task #%s deleted", taskID)}
	}

	// TS builds the updates object only from truthy fields (tools.ts:243-256);
	// empty strings are skipped, not applied.
	updates := map[string]any{}
	if hasStatus && status != "" {
		updates["status"] = status
	}
	if s, ok := args["subject"].(string); ok && s != "" {
		updates["subject"] = s
	}
	if s, ok := args["description"].(string); ok && s != "" {
		updates["description"] = s
	}
	if s, ok := args["owner"].(string); ok && s != "" {
		updates["owner"] = s
	}
	// TS validates both arrays with z.array(z.string()): a non-string element
	// fails the whole call (the Go wording stands in for the zod issue dump).
	for _, field := range []string{"addBlocks", "addBlockedBy"} {
		raw, present := args[field]
		if !present {
			continue
		}
		list, ok := raw.([]any)
		if !ok {
			return tools.ToolResult{
				Output:  fmt.Sprintf("Error: %s must be an array of strings", field),
				IsError: true,
			}
		}
		ids := make([]string, 0, len(list))
		for _, item := range list {
			s, ok := item.(string)
			if !ok {
				return tools.ToolResult{
					Output:  fmt.Sprintf("Error: %s must be an array of strings", field),
					IsError: true,
				}
			}
			ids = append(ids, s)
		}
		if len(ids) > 0 {
			updates[field] = raw
		}
	}
	// Go extension: metadata merge (null deletes a key).
	if m, ok := args["metadata"].(map[string]any); ok {
		updates["metadata"] = m
	}

	task, _, err := t.List.Update(taskID, updates)
	if err != nil {
		return tools.ToolResult{Output: fmt.Sprintf("Error: %s", err), IsError: true}
	}
	if task == nil {
		return tools.ToolResult{Output: "Task not found", IsError: true}
	}

	// TS always reports the same success line once the update call returns a
	// task, regardless of which fields actually changed (tools.ts:268).
	return tools.ToolResult{Output: fmt.Sprintf("Updated task #%s status", taskID)}
}
